src/HexagonOptimize.cpp

```diff
@@ -1079,11 +1081,64 @@ class OptimizePatterns : public IRMutator {
             // that they generate.
             internal_assert(op->args.size() == 3);
             return mutate(lower_lerp(op->args[0], op->args[1], op->args[2]));
+        } else if (op->is_intrinsic(Call::div_round_to_zero) &&
+                   !op->type.is_float() && op->type.is_vector()) {
+            internal_assert(op->args.size() == 2);
+            Expr a = op->args[0];
+            Expr b = op->args[1];
+            // Run bounds analysis to estimate the range of result.
+            Expr extent_upper = find_constant_bound(a / b, Direction::Upper, bounds);
+            const uint64_t *upper_bound = as_const_uint(extent_upper);
+
+            a = mutate(a);
+            b = mutate(b);
+            debug(1) << "Using long div: (num: " << a << "); (den: " << b << ")\n";
+            if (a.type().is_uint()) {
+                return unsigned_long_div(a, b, upper_bound);
+            }
+            // Use unsigned long div for signed integer division.
+            Type ty = a.type().with_code(Type::UInt);
+            Expr a_unsigned = cast(ty, abs(a));
+            Expr b_unsigned = cast(ty, abs(b));
+            Expr q = cast(a.type(),
+                          unsigned_long_div(a_unsigned, b_unsigned, upper_bound));
+            // Check the sign required for the quotient.
+            Expr a_neg = a >> make_const(a.type(), (a.type().bits() - 1));
+            Expr b_neg = b >> make_const(a.type(), (a.type().bits() - 1));
+            return q * ((a_neg ^ b_neg) | 1);
         } else {
             return IRMutator::visit(op);
         }
     }
 
+    template<typename NodeType, typename T>
+    NodeType visit_let(const T *op) {
+        // We only care about vector lets.
```

Comment:
Is this true? I think we need the bounds of scalars here, so we can understand the bounds of divisions with broadcasts in them?

Reply:
We do need to track scalars as well. Corrected it.